File: code_examples/zmq_server.js
var zmq = require("zeromq"),
  publisher = zmq.socket("pub"), // Vytvoření objektu publisher representujícího výstupní socket. Argument "pub" značí, že program bude zprávy do socketu posílat
  command_receiver = zmq.socket("sub"); // Vytvoření socketu command_receiver representujícího vstupní socket. Argument "sub" značí, že program bude zprávy ze socketu přijímat

publisher.bindSync("tcp://*:5558"); // Přiřazení adresy a portu socketu publisher

command_receiver.bindSync("tcp://*:5559"); // Přiřazení adresy a portu socketu command_receiver
command_receiver.subscribe(""); // Přihlášení k odběru všech zpráv

command_receiver.on("message", (msg) => { // Při přijetí zprávy
  console.log("Přijatá zpráva: ", msg.toString()); // Vypsání přijaté zprávy do konzole
  publisher.send("prijal jsem zpravu: " + msg.toString()); // Odeslání zprávy do socketu publisher
});

File: code_examples/zmq_server.cpp
//  Clone server Model One

#include "zmq.hpp"
#include "zmq_addon.hpp"
#include <iostream>
#include <string>

int main(void)
{
    zmq::context_t ctx(1); // Třída context_t je používána k vytvoření všech soketů v programu.

    zmq::socket_t publisher(ctx, zmq::socket_type::pub); // Vytvoření třídy socket_t reprezentující výstupní socket. Druhý argument určuje, zda do socketu můžeme zapisovat (publish) nebo z něj číst (subscribe).
    publisher.bind("tcp://*:5556"); // Funkcí bind je socketu přiřazena adresa. V tomto případě socket pracuje s protokolem tcp na portu 5556.
    
    zmq::socket_t command_receiver(ctx, zmq::socket_type::sub); // Vytvoření třídy reprezentující vstupní socket.
    command_receiver.bind("tcp://*:5557"); // Vstupnímu socketu je přiřazena adresa.
    command_receiver.set(zmq::sockopt::subscribe, ""); // Jestliže je socket typu sub, je třeba nastavit, na která témata bude reagovat. V tomto případě je nastaveno, že bude reagovat na všechna témata.

    while (true)
    {
        zmq::message_t received; // Vytvoření třídy message_t, která reprezentuje zprávu. Při príjmání zpráv je třeba mít předem vytvořenou třídu message_t, do které se zpráva uloží.
        if (command_receiver.recv(received, zmq::recv_flags::none)) // Příjem zprávy. Příjem je blokující, dokud není zpráva přijata. Při použití argumentu zmq::recv_flags::dontwait by byl příjem neblokující. Funkce vrací true, pokud byla zpráva úspěšně přijata, jinak vrací false.
        {
        std::cout << "přijato: \"" << received.to_string() << "\"" << std::endl;

        std::string string_to_be_sent = "potvrzuji, přijal jsem zprávu \"" + received.to_string() + "\"";

        zmq::message_t to_be_sent(string_to_be_sent.c_str(), string_to_be_sent.length() + 1); //Vytvoření zprávy, která bude odeslána.
        publisher.send(to_be_sent, zmq::send_flags::none); // Odeslání zprávy.
        }
        else {
            std::cout << "Nebyla úspěšně přijata žádná zpráva."
        }

    }
    return 0;
}